simpleclient_pushgateway/src/main/java/io/prometheus/client/exporter/PushGateway.java

```diff
@@ -54,15 +56,42 @@
  */
 public class PushGateway {
 
-  private final String address;
+  // Visible for testing
+  final String gatewayBaseURL;
   private static final int MILLISECONDS_PER_SECOND = 1000;
   /**
    * Construct a Pushgateway, with the given address.
    * <p>
    * @param address  host:port or ip:port of the Pushgateway.
    */
   public PushGateway(String address) {
-    this.address = address;
+    this(createURLSneakily("http://" + address));
+  }
+
+  /**
+   * Construct a Pushgateway, with the given URL.
+   * <p>
+   * @param serverBaseURL the base URL and optional context path of the Pushgateway server.
+   */
+  public PushGateway(URL serverBaseURL) {
+    this.gatewayBaseURL = URI.create(serverBaseURL.toString() + "/metrics/job/")
+      .normalize()
+      .toString();
```

Comment:
After trying a lot of different approaches to getting a normalised URL as a String out of this constructor, I ended up here.

The constructor accepts a `URL`, so we know that it has the required URL properties (i.e., a scheme is present and `:` is followed by either a valid port number or `/`).

Once we've established those facts, there is no harm in turning it into a more general `URI` for normalisation. I'm more happy with this solution.

Side note: I'm surprised at how loose the `URL` class's validator is. For example, `http://http://` is considered to be a valid URL.